libp2p/stream_muxer/yamux/yamux.py

```diff
@@ -0,0 +1,270 @@
+"""
+Yamux stream multiplexer implementation for py-libp2p.
+This is the preferred multiplexing protocol due to its performance and feature set.
+Mplex is also available for legacy compatibility but may be deprecated in the future.
+"""
+import logging
+import struct
+from typing import (
+    Optional,
+)
+
+import trio
+from trio import (
+    MemoryReceiveChannel,
+    MemorySendChannel,
+    Nursery,
+)
+
+from libp2p.abc import (
+    IMuxedConn,
+    IMuxedStream,
+    ISecureConn,
+)
+from libp2p.peer.id import (
+    ID,
+)
+from libp2p.stream_muxer.exceptions import (
+    MuxedStreamError,
+)
+
+PROTOCOL_ID = "/yamux/1.0.0"
+TYPE_DATA = 0x0
+TYPE_WINDOW_UPDATE = 0x1
+TYPE_PING = 0x2
+TYPE_GO_AWAY = 0x3
+FLAG_SYN = 0x1
+FLAG_ACK = 0x2
+FLAG_FIN = 0x4
+FLAG_RST = 0x8
+HEADER_SIZE = 12
+
+
+class YamuxStream(IMuxedStream):
+    def __init__(self, stream_id: int, conn: "Yamux", is_initiator: bool) -> None:
+        self.stream_id = stream_id
+        self.conn = conn
+        self.is_initiator = is_initiator
+        self.closed = False
+
+    async def write(self, data: bytes) -> None:
+        if self.closed:
+            raise MuxedStreamError("Stream is closed")
+        header = struct.pack("!BBHII", 0, TYPE_DATA, 0, self.stream_id, len(data))
+        await self.conn.secured_conn.write(header + data)
+
+    async def read(self, n: int = -1) -> bytes:
+        return await self.conn.read_stream(self.stream_id, n)
+
+    async def close(self) -> None:
+        if not self.closed:
+            logging.debug(f"Closing stream {self.stream_id}")
+            header = struct.pack("!BBHII", 0, TYPE_DATA, FLAG_FIN, self.stream_id, 0)
+            await self.conn.secured_conn.write(header)
+            self.closed = True
+
+    async def reset(self) -> None:
+        if not self.closed:
+            logging.debug(f"Resetting stream {self.stream_id}")
+            header = struct.pack("!BBHII", 0, TYPE_DATA, FLAG_RST, self.stream_id, 0)
+            await self.conn.secured_conn.write(header)
+            self.closed = True
+
+    def set_deadline(self, ttl: int) -> bool:
```

Comment:
What does this do, that it should always return `False`?

Reply:
The `set_deadline` method returns `False` because Yamux doesn’t natively support setting deadlines (unlike Mplex, which tracks them). I’ve added a docstring to clarify this: "Yamux does not support deadlines natively, so this method always returns False." If we need deadline support, we could extend it later, but for now, this aligns with the spec. Thoughts?

Reply:
The comment is good, but for an end user, I think raising a `NotImplementedError` is probably more useful.

Reply:
Alright, I’ve implemented the change to raise `NotImplementedError` in `set_deadline`. I hit an `AttributeError: 'YamuxStream' object has no attribute 'muxed_conn'` when adding a test for `set_deadline`, due to `NetStream` expecting `muxed_conn`. I fixed it by adding `self.muxed_conn = conn` to `YamuxStream.__init__`, consistent with `MplexStream`. I verify it raises `NotImplementedError`. All `yamux` and `mplex` tests now pass.